Panic on resource import (deposition in Zenodo API)

Hi! Thanks for the lovely provider, @magodo!

I'm trying to import an existing deposition resource via the Zenodo REST API (see https://developers.zenodo.org/#retrieve for documentation).

Here is a minimal configuration file to reproduce the error:

```terraform
terraform {
  required_providers {
    restful = {
      source  = "magodo/restful"
      version = "0.22.0"
    }
  }
}

provider "restful" {
  base_url = "https://sandbox.zenodo.org"
  security = {
    apikey = [
      {
        in    = "header"
        name  = "Authorization"
        value = "Bearer XXX" # NOTE: you need to register at https://sandbox.zenodo.org and generate a token
      },
    ]
  }
  header = {
    "Content-Type" = "application/json"
  }
}

import {
  to = restful_resource.test_deposition
  id = jsonencode({
    path = "/api/deposit/depositions/274856" # NOTE: you'll want to replace occurrences of 274856 with the id of a deposition created by your user
    id   = "/api/deposit/depositions/274856"
  })
}

resource "restful_resource" "test_deposition" {
  path = "/api/deposit/depositions/274856"
  body = { metadata = {} }
}
```

LMK if i can get you via a private channel the auth token to apply the configuration above.


Here is the stack trace of the error:

```
❯ terraform plan
restful_resource.test_deposition: Preparing import... [id={"id":"/api/deposit/depositions/274856","path":"/api/deposit/depositions/274856"}]
restful_resource.test_deposition: Refreshing state... [id=/api/deposit/depositions/274856]

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Request cancelled
│
│ The plugin6.(*GRPCProvider).ReadResource request was cancelled.
╵

Stack trace from the terraform-provider-restful_v0.22.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x40 pc=0x1024ef710]

goroutine 25 [running]:
github.com/magodo/terraform-provider-restful/internal/provider.Resource.read({0x2?}, {0x102b56a38, 0x14000010f30}, {{{{0x102b5bb18, 0x140005b8b10}, {0x102a3c920, 0x140005b8000}}, {0x102b5f528, 0x140005846e0}}, 0x140003b8000, ...}, ...)
        github.com/magodo/terraform-provider-restful/internal/provider/resource.go:1039 +0xdc0
github.com/magodo/terraform-provider-restful/internal/provider.Resource.Read({0x0?}, {0x102b56a38, 0x14000010f30}, {{{{0x102b5bb18, 0x140005b8b10}, {0x102a3c920, 0x140005b8000}}, {0x102b5f528, 0x140005846e0}}, 0x140003b8000, ...}, ...)
        github.com/magodo/terraform-provider-restful/internal/provider/resource.go:917 +0xb4
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ReadResource(0x1400022a908, {0x102b56a38, 0x14000010f30}, 0x14000010f90, 0x140003e55d8)
        github.com/hashicorp/terraform-plugin-framework@v1.14.1-0.20250219211921-3000d8cd3b43/internal/fwserver/server_readresource.go:119 +0x688
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ReadResource(0x1400022a908, {0x102b56a38?, 0x14000010e40?}, 0x14000550400)
        github.com/hashicorp/terraform-plugin-framework@v1.14.1-0.20250219211921-3000d8cd3b43/internal/proto6server/server_readresource.go:55 +0x2dc
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ReadResource(0x140002661e0, {0x102b56a38?, 0x14000010330?}, 0x140003604d0)
        github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov6/tf6server/server.go:785 +0x210
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ReadResource_Handler({0x102b26560, 0x140002661e0}, {0x102b56a38, 0x14000010330}, 0x140005a8580, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:575 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000242200, {0x102b56a38, 0x140000102d0}, 0x1400035d200, 0x14000287da0, 0x10314b988, 0x0)
        google.golang.org/grpc@v1.69.4/server.go:1392 +0xc38
google.golang.org/grpc.(*Server).handleStream(0x14000242200, {0x102b57558, 0x1400047c000}, 0x1400035d200)
        google.golang.org/grpc@v1.69.4/server.go:1802 +0x900
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.69.4/server.go:1030 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 39
        google.golang.org/grpc@v1.69.4/server.go:1041 +0x138

Error: The terraform-provider-restful_v0.22.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

Since registering to the Zenodo sandbox might me a nuisance for you, here's the call i'd expect the provider to be doing:

```
❯ curl -i -X GET 'https://sandbox.zenodo.org/api/deposit/depositions/274856?access_token=XXX'
HTTP/1.1 200 OK
server: nginx
date: Mon, 23 Jun 2025 20:11:01 GMT
content-type: application/json
content-length: 2295
vary: Accept-Encoding
etag: "56"
x-ratelimit-limit: 1000
x-ratelimit-remaining: 998
x-ratelimit-reset: 1750709513
retry-after: 51
permissions-policy: interest-cohort=()
x-frame-options: sameorigin
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
content-security-policy: default-src 'self' fonts.googleapis.com *.gstatic.com data: 'unsafe-inline' 'unsafe-eval' blob: zenodo-broker.web.cern.ch zenodo-broker-qa.web.cern.ch maxcdn.bootstrapcdn.com cdnjs.cloudflare.com ajax.googleapis.com webanalytics.web.cern.ch
strict-transport-security: max-age=31556926; includeSubDomains
referrer-policy: strict-origin-when-cross-origin
access-control-allow-origin: *
access-control-expose-headers: Content-Type, ETag, Link, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset
strict-transport-security: max-age=15768000
x-request-id: 81355b2633c1dcb4686f5bc9a1fed6dd
set-cookie: 04f20c86f07421a9ec0f9d5ba4be544f=e2703e3fe7a463a9f39917b85618cae5; path=/; HttpOnly; Secure; SameSite=None
cache-control: private

{"created": "2025-06-21T11:17:59.706197+00:00", "modified": "2025-06-21T11:29:29.629467+00:00", "id": 274856, "conceptrecid": "274855", "doi": "10.5072/zenodo.274856", "conceptdoi": "10.5072/zenodo.274855", "doi_url": "https://doi.org/10.5072/zenodo.274856", "metadata": {"title": "Gigatest", "doi": "10.5072/zenodo.274856", "publication_date": "2025-06-25", "description": "This is description.", "access_right": "open", "creators": [{"name": "Smith, Jane", "affiliation": null, "orcid": "0000-0002-1694-233X"}], "license": "cc-zero", "imprint_publisher": "Zenodo", "upload_type": "dataset", "prereserve_doi": {"doi": "10.5281/zenodo.274856", "recid": 274856}}, "title": "Gigatest", "links": {"self": "https://sandbox.zenodo.org/api/deposit/depositions/274856", "html": "https://sandbox.zenodo.org/deposit/274856", "doi": "https://doi.org/10.5072/zenodo.274856", "badge": "https://sandbox.zenodo.org/badge/doi/10.5072%2Fzenodo.274856.svg", "files": "https://sandbox.zenodo.org/api/deposit/depositions/274856/files", "bucket": "https://sandbox.zenodo.org/api/files/84885422-df68-47ee-ae6f-35244999e3d9", "latest_draft": "https://sandbox.zenodo.org/api/deposit/depositions/274856", "latest_draft_html": "https://sandbox.zenodo.org/deposit/274856", "publish": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/publish", "edit": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/edit", "discard": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/discard", "newversion": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/newversion", "record": "https://sandbox.zenodo.org/api/records/274856", "record_html": "https://sandbox.zenodo.org/record/274856", "latest": "https://sandbox.zenodo.org/api/records/274856/versions/latest", "latest_html": "https://sandbox.zenodo.org/records/274856/latest"}, "record_id": 274856, "owner": 44770, "files": [{"id": "04088627-5c74-4b79-a40b-874acc782ae5", "filename": ".keepme", "filesize": 8, "checksum": "da385ad1b614e80e93b37c771e979a16", "links": {"self": "https://sandbox.zenodo.org/api/deposit/depositions/274856/files/04088627-5c74-4b79-a40b-874acc782ae5", "download": "https://sandbox.zenodo.org/api/records/274856/draft/files/.keepme/content"}}], "state": "inprogress", "submitted": true}%
```

Here's the response payload, prettified:

```json
{
  "created": "2025-06-21T11:17:59.706197+00:00",
  "modified": "2025-06-21T11:29:29.629467+00:00",
  "id": 274856,
  "conceptrecid": "274855",
  "doi": "10.5072/zenodo.274856",
  "conceptdoi": "10.5072/zenodo.274855",
  "doi_url": "https://doi.org/10.5072/zenodo.274856",
  "metadata": {
    "title": "Gigatest",
    "doi": "10.5072/zenodo.274856",
    "publication_date": "2025-06-25",
    "description": "This is description.",
    "access_right": "open",
    "creators": [
      {
        "name": "Smith, Jane",
        "affiliation": null,
        "orcid": "0000-0002-1694-233X"
      }
    ],
    "license": "cc-zero",
    "imprint_publisher": "Zenodo",
    "upload_type": "dataset",
    "prereserve_doi": {
      "doi": "10.5281/zenodo.274856",
      "recid": 274856
    }
  },
  "title": "Gigatest",
  "links": {
    "self": "https://sandbox.zenodo.org/api/deposit/depositions/274856",
    "html": "https://sandbox.zenodo.org/deposit/274856",
    "doi": "https://doi.org/10.5072/zenodo.274856",
    "badge": "https://sandbox.zenodo.org/badge/doi/10.5072%2Fzenodo.274856.svg",
    "files": "https://sandbox.zenodo.org/api/deposit/depositions/274856/files",
    "bucket": "https://sandbox.zenodo.org/api/files/84885422-df68-47ee-ae6f-35244999e3d9",
    "latest_draft": "https://sandbox.zenodo.org/api/deposit/depositions/274856",
    "latest_draft_html": "https://sandbox.zenodo.org/deposit/274856",
    "publish": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/publish",
    "edit": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/edit",
    "discard": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/discard",
    "newversion": "https://sandbox.zenodo.org/api/deposit/depositions/274856/actions/newversion",
    "record": "https://sandbox.zenodo.org/api/records/274856",
    "record_html": "https://sandbox.zenodo.org/record/274856",
    "latest": "https://sandbox.zenodo.org/api/records/274856/versions/latest",
    "latest_html": "https://sandbox.zenodo.org/records/274856/latest"
  },
  "record_id": 274856,
  "owner": 44770,
  "files": [
    {
      "id": "04088627-5c74-4b79-a40b-874acc782ae5",
      "filename": ".keepme",
      "filesize": 8,
      "checksum": "da385ad1b614e80e93b37c771e979a16",
      "links": {
        "self": "https://sandbox.zenodo.org/api/deposit/depositions/274856/files/04088627-5c74-4b79-a40b-874acc782ae5",
        "download": "https://sandbox.zenodo.org/api/records/274856/draft/files/.keepme/content"
      }
    }
  ],
  "state": "inprogress",
  "submitted": true
}
```

--- a/internal/provider/resource.go
+++ b/internal/provider/resource.go
@@ -1036,7 +1036,12 @@ func (r Resource) read(ctx context.Context, req resource.ReadRequest, resp *reso
 		}
 
 		var body types.Dynamic
-		if body, err = dynamic.FromJSON(b, state.Body.UnderlyingValue().Type(ctx)); err != nil {
+		if state.Body.IsNull() {
+			body, err = dynamic.FromJSONImplied(b)
+		} else {
+			body, err = dynamic.FromJSON(b, state.Body.UnderlyingValue().Type(ctx))
+		}
+		if err != nil {
 			// An error might occur here during refresh, when the type of the state doesn't match the remote,
 			// e.g. a tuple field has different number of elements.
 			// In this case, we fallback to the implied types, to make the refresh proceed and return a reasonable plan diff.

--- a/internal/provider/resource_jsonserver_test.go
+++ b/internal/provider/resource_jsonserver_test.go
@@ -68,6 +68,32 @@ func TestResource_JSONServer_Basic(t *testing.T) {
 					return fmt.Sprintf(`{"id": %q, "path": "posts", "body": {"foo": null}}`, s.RootModule().Resources[addr].Primary.Attributes["id"]), nil
 				},
 			},
+			// Should also be able to import without "body" specified.
+			{
+				ResourceName:            addr,
+				ImportState:             true,
+				ImportStateVerify:       false,
+				ImportStateVerifyIgnore: []string{"read_path"},
+				ImportStateIdFunc: func(s *terraform.State) (string, error) {
+					return fmt.Sprintf(`{"id": %q, "path": "posts"}`, s.RootModule().Resources[addr].Primary.Attributes["id"]), nil
+				},
+				ImportStateCheck: func(is []*terraform.InstanceState) error {
+					if len(is) != 1 {
+						return fmt.Errorf("expect instance state has only one item, got=%d", len(is))
+					}
+					inst := is[0]
+					if v := inst.Attributes["body.%"]; v != "2" {
+						return fmt.Errorf(`expect "body.%%" to be "2", got=%s`, v)
+					}
+					if v := inst.Attributes["body.foo"]; v != "foo" {
+						return fmt.Errorf(`expect "body.foo" to be "foo", got=%s`, v)
+					}
+					if _, ok := inst.Attributes["body.id"]; !ok {
+						return fmt.Errorf(`expect "body.id" to exist`)
+					}
+					return nil
+				},
+			},
 			{
 				Config: d.basic("bar"),
 				ConfigStateChecks: []statecheck.StateCheck{

